Design/Packaging/Future_packaging_strategy.md

```diff
@@ -0,0 +1,70 @@
+<!-- This document's diagrams are produced using the .plantuml documents in this file's directory and plantuml -->
+# Future packaging with conda
+## Current methodology
+At present we have bespoke methodologies for packaging on each different OS we support (Windows, MacOSX, Ubuntu, and CentOS/RHEL7).
+
+- On Windows, we create a packaged installer containing our dependencies and our compiled program as a .exe. We only link to some parts of System32 that provide Windows functionality required for Mantid, and the OpenGL shim to handle the graphics drivers. We install Mantid running the .exe installer by double clicking on the executable.
+
+- On MacOSX, we create a package containing our dependencies and our compiled program in a .dmg app bundle. Only low level packages that are expected to exist on MacOSX. We install Mantid .dmg by running the installer by double clicking on the bundle.
+
+- On Ubuntu, we create a package containing only our compiled program in a .deb format. We rely on the the currently available system packages to run Mantid. We install the Mantid .deb package using dpkg.
+
+- On CentOS/RHEL7, we create a package containing only our compiled program in a .rpm format. We rely on the currently available system packages, optional packages from `rhel-7-workstation-optional-rpms` and some packages from EPEL to run Mantid. We install the Mantid .rpm package using the rpm command.
+
+## Intended move forward
+
+All operating systems should have a fully built and deployable conda-build implementation, where a use can in theory run a simple command in conda and it will install Mantid with all it's runtime dependencies, which with some small tweaks we can then package on all Operating Systems assuming that the conda-build recipe has been created and ran successfully.
+
+The aim at this time is to create similar packages to what we have now on MacOSX and Windows, whilst shipping all dependencies in our Ubuntu and CentOS/RHEL7 which is a change. We will make a .exe for Windows, .dmg for MacOSX, .deb for Ubuntu, and a .rpm for CentOS/RHEL7 much like we do now on all operating systems.
+
+### Conda build
+
+We intend to distribute a conda package using the conda-build system, which will allow advanced users to install Mantid using conda. Users will be able to install this packaged version with, 
```

Comment:
Can we include here that we intend to build separate conda packages for::
 - `mantid-framework` (as we already do)
 - `mantid-qt`
 - `mantid-workbench`

The first two are libraries that other applications can reuse.

Reply:
Another thing to potentially consider here (something we ran into with scipp, and haven't fully decided on a solution yet):

How should optional dependencies be handled? Not sure what a good example of this would be in Mantid, but I am thinking of things that may be required to use certain features (in scipp plotting is an example: You don't really need matplotlib if you just want to use the lib, but if you want to plot, you do). Would you create meta packages such as `mantid-minimal` and `mantid-all`?

Reply:
> Can we include here that we intend to build separate conda packages for::
> 
>     * `mantid-framework` (as we already do)
> 
>     * `mantid-qt`
> 
>     * `mantid-workbench`
> 
> 
> The first two are libraries that other applications can reuse.

I didn't know we were looking at producing a `mantid-qt` package, perhaps we should discuss scope/whether or not this is expected as part of the current effort towards conda-ification.

Reply:
> Another thing to potentially consider here (something we ran into with scipp, and haven't fully decided on a solution yet):
> 
> How should optional dependencies be handled? Not sure what a good example of this would be in Mantid, but I am thinking of things that may be required to use certain features (in scipp plotting is an example: You don't really need matplotlib if you just want to use the lib, but if you want to plot, you do). Would you create meta packages such as `mantid-minimal` and `mantid-all`?

I think I discussed this briefly in the Future looking part of the document. For now, my understanding is that we do not want to support optional dependencies, as for most of the dependencies they are absolutely required. Right now we are producing a `mantid-all` effectively. I think long term it is very possible to make `mantid-minimal` through to some midpoint for example `mantid-muon` to `mantid-all` where only the absolutely needed dependencies are pulled in.

Reply:
> > Can we include here that we intend to build separate conda packages for::
> > ```
> > * `mantid-framework` (as we already do)
> > 
> > * `mantid-qt`
> > 
> > * `mantid-workbench`
> > ```
> > The first two are libraries that other applications can reuse.
> 
> I didn't know we were looking at producing a `mantid-qt` package, perhaps we should discuss scope/whether or not this is expected as part of the current effort towards conda-ification.

Things like [mslice](https://github.com/mantidproject/mslice), [PyRs](https://github.com/neutrons/PyRs), [addie](https://github.com/neutrons/addie) use features from the `mantidqt` library but are separate and to do not need workbench so it is beneficial to have them separated.

Reply:
> Another thing to potentially consider here (something we ran into with scipp, and haven't fully decided on a solution yet):
> 
> How should optional dependencies be handled? Not sure what a good example of this would be in Mantid, but I am thinking of things that may be required to use certain features (in scipp plotting is an example: You don't really need matplotlib if you just want to use the lib, but if you want to plot, you do). Would you create meta packages such as `mantid-minimal` and `mantid-all`?

We don't have a good solution to this either to be honest. We have a few algorithms that can use dependencies if they are available and it would be nice to to have to include everything as a hard dependency. We probably have a worse problem than `scipp` in terms of the number of ways we can carve up the codebase so what future meta-packages would be is challenging. Were you thinking of going to the level of a `scipp-core` and then separate facility packages?

Reply:
Since you last saw `scipp` we have actually removed anything neutron-specific form it, so what I am mentioning above is even more generic dependencies. There is now a separate `scippneutron`. The latter does (and will not) not contain facility-specifics either, those have to be in separate packages (we have `ess` for this).

Reply:
We have packages that would use `mantid-qt`, but could require `mantid-workbench` instead. That is essentially what we did long ago.

Reply:
In response, @StephenSmith25 has put together a plan to move towards `MantidQt` being independent of workbench, and dependent on `MantidFramework` in order to compile from `MantidFramework` some work will need to be done to ensure that this package provides the correct headers. Then the end result of all of this a `MantidWorkbench` package will be produced that is dependent on `MantidFramework` and `MantidQt` and will depend on the available headers that these provide at compile time for Conda Build. This is my understanding of how work is progressing at present, I will put something in the document that speaks to this effect.